### PR TITLE
Add OptimizeContext MSBuild target that runs on build

### LINKED ISSUE #24894: Generate the compiled model during build
This allows to make sure that the context is using an up-to-date model, most useful for Release builds.
To avoid dependency cycles the compiled model needs to be compiled in an assembly separate from the startup assembly

### PR DESCRIPTION
**Pull request by AndriySvyryd**
Tests tracked by https://github.com/dotnet/efcore/issues/33136

Even though the compiled model is added to the project, it's not yet discovered, this is tracked by #24893

Fixes #24894

### CODE REVIEW

**Review by roji — COMMENTED** [review]
@AndriySvyryd is the idea here to generate the compiled model on every build (during the inner loop), as opposed to on publish? Assuming that we don't have the ability to know when something has changed that requires a model rebuild, is this something we want to do, given that it will make the build much slower?

**Comment by AndriySvyryd** [comment]
> @AndriySvyryd is the idea here to generate the compiled model on every build (during the inner loop), as opposed to on publish?

The idea here is to have a task to generate the compiled model when we decide it should be generated.

> given that it will make the build much slower?

We need to measure it first

**Comment by roji** [comment]
> The idea here is to have a task to generate the compiled model when we decide it should be generated.

OK, sure thing.

File: src/EFCore.Tasks/Tasks/Internal/MsBuildUtilities.cs [hunk 1]
```diff
@@ -0,0 +1,85 @@
...
+///     This is an internal API that supports the Entity Framework Core infrastructure and not subject to
+///     the same compatibility standards as public APIs. It may be changed or removed without notice in
+///     any release. You should only use it directly in your code with extreme caution and knowing that
+///     doing so can result in application failures when updating to a new Entity Framework Core release.
+/// </summary>
+internal class MsBuildUtilities
+{
+    public static string[] Split(string s)
+    {
+        if (!string.IsNullOrEmpty(s))
```
Comment (roji) [inline src/EFCore.Tasks/Tasks/Internal/MsBuildUtilities.cs:20]: nit: conditional operators and expression-bodied methods throughout this file?

File: src/EFCore.Tasks/Tasks/Internal/MsBuildUtilities.cs [hunk 2]
```diff
@@ -0,0 +1,85 @@
...
+        {
+            return s.Split(';')
+                .Select(entry => entry.Trim())
+                .Where(entry => entry.Length != 0)
+                .ToArray();
+        }
+        return Array.Empty<string>();
```
Comment (roji) [inline src/EFCore.Tasks/Tasks/Internal/MsBuildUtilities.cs:27]: nit

```suggestion
        return [];
```

**Review by roji — COMMENTED** [review]
Just some nits

File: src/EFCore.Tasks/EFCore.Tasks.csproj [hunk 3]
```diff
@@ -0,0 +1,88 @@
...
+    <ItemGroup>
+      <NuspecProperty Include="AssemblyName=$(AssemblyName)" />
+      <NuspecProperty Include="OutputPath=$(OutputPath)" />
+      <NuspecProperty Include="DefaultNetCoreTargetFramework=$(DefaultNetCoreTargetFramework)" />
+      <NuspecProperty Include="Configuration=$(Configuration)" />
+    </ItemGroup>
+  </Target>
+
+  <ItemGroup>
+    <Service Include="{508349b6-6b84-4df5-91f0-309beebad82d}" />
```
Comment (SamMonoRT) [inline src/EFCore.Tasks/EFCore.Tasks.csproj:85]: For my knowledge - where is this ID coming from?

File: src/EFCore.Tasks/EFCore.Tasks.csproj [hunk 4]
```diff
@@ -0,0 +1,88 @@
...
+  </ItemGroup>
+
+  <ItemGroup Condition="'$(TargetFramework)' != 'net472'">
+    <PackageReference Include="Microsoft.Build.Framework" Version="$(MicrosoftBuildFrameworkVersion)" PrivateAssets="all" ExcludeAssets="Runtime" IsImplicitlyDefined="true" />
+    <PackageReference Include="Microsoft.Build.Tasks.Core" Version="$(MicrosoftBuildTasksCoreVersion)" PrivateAssets="all" ExcludeAssets="Runtime" IsImplicitlyDefined="true" />
+    <PackageReference Include="Microsoft.Build.Utilities.Core" Version="$(MicrosoftBuildUtilitiesCoreVersion)" PrivateAssets="all" ExcludeAssets="Runtime" IsImplicitlyDefined="true" />
+  </ItemGroup>
+
+  <ItemGroup Condition="'$(TargetFramework)' == 'net472'">
+    <PackageReference Include="System.Text.Json" Version="$(SystemTextJsonVersion)" PrivateAssets="All" />
```
Comment (SamMonoRT) [inline src/EFCore.Tasks/EFCore.Tasks.csproj:57]: nit- Sort in alphabetical order

File: eng/Versions.props [hunk 5]
```diff
@@ -31,10 +31,13 @@
     <MicrosoftDotNetBuildTasksTemplatingVersion>9.0.0-beta.23607.2</MicrosoftDotNetBuildTasksTemplatingVersion>
   </PropertyGroup>
   <PropertyGroup Label="Other dependencies">
+    <MicrosoftBuildFrameworkVersion>17.0.0</MicrosoftBuildFrameworkVersion>
```
Comment (SamMonoRT) [inline eng/Versions.props:34]: Q for my knowledge - Do we know from past history, how often do we need to update this version #s. What happens when this changes and we don't update it here?

File: src/ef/Reporter.cs [hunk 6]
```diff
@@ -8,6 +8,12 @@ namespace Microsoft.EntityFrameworkCore.Tools;
 
 internal static class Reporter
 {
+    public const string ErrorPrefix = "error:   ";
```
Comment (SamMonoRT) [inline src/ef/Reporter.cs:11]: Q-are the extra spaces afte error (and others) only to make it more readable/parsing?

**Review by SamMonoRT — APPROVED** [review]

File: eng/Versions.props (hunk 5 above)
Comment (AndriySvyryd) [reply eng/Versions.props:34]: Generally, we shouldn't need to update this unless we need to use a new feature or there's been a breaking change in MsBuild

File: src/ef/Reporter.cs (hunk 6 above)
Comment (AndriySvyryd) [reply src/ef/Reporter.cs:11]: They are for readability, so different kinds of messages align vertically

**Comment by cincuranet** [comment]
@JanKrivanek PTAL, thanks.

File: src/EFCore.Tasks/EFCore.Tasks.nuspec [hunk 7]
```diff
@@ -0,0 +1,23 @@
+<?xml version="1.0" encoding="utf-8"?>
+
+<package xmlns="http://schemas.microsoft.com/packaging/2013/05/nuspec.xsd">
+  <metadata>
+    $CommonMetadataElements$
+    <dependencies>
+      <group targetFramework=".NET8.0">
```
Comment (JanKrivanek) [inline src/EFCore.Tasks/EFCore.Tasks.nuspec:7]: Should this be `$(DefaultNetCoreTargetFramework)`?

File: src/EFCore.Tasks/EFCore.Tasks.nuspec [hunk 8]
```diff
@@ -0,0 +1,23 @@
...
+    <readme>docs\PACKAGE.md</readme>
+  </metadata>
+  <files>
+    $CommonFileElements$
+    <file src="PACKAGE.md" target="docs\" />
+    <file src="$OutputPath$$DefaultNetCoreTargetFramework$\$AssemblyName$.dll" target="tasks\$DefaultNetCoreTargetFramework$\" />
+    <file src="$OutputPath$$DefaultNetCoreTargetFramework$\$AssemblyName$.pdb" target="tasks\$DefaultNetCoreTargetFramework$\" />
+    <file src="$OutputPath$$DefaultNetCoreTargetFramework$\$AssemblyName$.deps.json" target="tasks\$DefaultNetCoreTargetFramework$\" />
+    <file src="$OutputPath$net472\*" target="tasks\net472\" />
+    <file src="..\..\artifacts\bin\ef\$Configuration$\netcoreapp2.0\*" target="tools\netcoreapp2.0\" />
```
Comment (JanKrivanek) [inline src/EFCore.Tasks/EFCore.Tasks.nuspec:20]:
```
    <file src="$(ArtifactsBinDir)ef\$Configuration$\netcoreapp2.0\*" target="tools\netcoreapp2.0\" />
```

To be a future proof against moving the file. If you do not want to be dependent on arcade - then you can use your $(SolutionRoot) property.

File: src/EFCore.Tasks/Tasks/Internal/OperationTaskBase.cs [hunk 9]
```diff
@@ -0,0 +1,253 @@
...
+
+namespace Microsoft.EntityFrameworkCore.Tasks.Internal;
+
+/// <summary>
+///     This is an internal API that supports the Entity Framework Core infrastructure and not subject to
+///     the same compatibility standards as public APIs. It may be changed or removed without notice in
+///     any release. You should only use it directly in your code with extreme caution and knowing that
+///     doing so can result in application failures when updating to a new Entity Framework Core release.
+/// </summary>
+public abstract class OperationTaskBase : Build.Utilities.Task
```
Comment (JanKrivanek) [inline src/EFCore.Tasks/Tasks/Internal/OperationTaskBase.cs:19]: It looks like [ToolTask](https://learn.microsoft.com/en-us/dotnet/api/microsoft.build.utilities.tooltask?view=msbuild-17-netcore) might be better fit here?

**Review by JanKrivanek — COMMENTED** [review]
@cincuranet - no strong concerns from msbuild point of view.

**Pull request merged**